File: python/Lec 2 Build-up Logical Thinking (Patterns)/pattern_03.py
# 1
# 12
# 123
# 1234
# 12345

#----------------------------------
# BRUTE FORCE APPROACH
#----------------------------------
n = 5
for i in range(1, n + 1):
    line = ""
    for j in range(1, i + 1):
        line += str(j)  # O(n²) string copying
    print(line)

#----------------------------------
# BETTER APPROACH (Using list and join)
#----------------------------------
n = 5
for i in range(1, n + 1):
    line = ''.join(str(j) for j in range(1, i + 1))
    print(line)

#----------------------------------
# BEST APPROACH (Single string accumulation)
#----------------------------------
n = 5
s = ""
for i in range(1, n + 1):
    s += str(i)  # Amortized O(1) appends
    print(s)

File: cpp/Lec 2 Build-up Logical Thinking (Patterns)/pattern_03.cpp
// 1
// 12
// 123
// 1234
// 12345

//----------------------------------
// BRUTE FORCE APPROACH
//----------------------------------
#include <iostream>
using namespace std;
int main()
{
    int n = 5;
    for (int i = 1; i <= n; i++)
    {
        for (int j = 1; j <= i; j++)
        {
            cout << j;
        }
        cout << endl;
    }
    return 0;
}

//----------------------------------
// BETTER APPROACH (Using string building)
//----------------------------------
#include <iostream>
#include <string>
using namespace std;
int main()
{
    int n = 5;
    for (int i = 1; i <= n; i++)
    {
        string s = "";
        for (int j = 1; j <= i; j++)
        {
            s += to_string(j);
        }
        cout << s << endl;
    }
    return 0;
}

//----------------------------------
// BEST APPROACH (Single pass with pre-allocation)
//----------------------------------
#include <iostream>
#include <string>
using namespace std;
int main()
{
    int n = 5;
    string s;
    s.reserve(15); // Pre-allocate memory
    for (int i = 1; i <= n; i++)
    {
        s += to_string(i);
        cout << s << '\n';
    }
    return 0;
}